### PR TITLE
[major] Support new authentication for File Catalog

### PR DESCRIPTION
**Pull request by blinkdog**
Work in progress

### CODE REVIEW

**Comment by blinkdog** [comment]
As I work through a bit of this, I'll probably ping you a few times @ric-evans

**Comment by ric-evans** [comment]
I think this is yours: https://github.com/WIPACrepo/file-catalog-indexer/issues/53 :smile: 

But I'm here for assistance!

**Comment by blinkdog** [comment]
Okay @ric-evans , I think we're ready for a review on this set of changes.

It doesn't include any changes for IceProd yet, I need to confirm with @dsschult if IceProd clients are using the `ClientCredentialsAuth` scheme or a different scheme.

**Review by dsschult — COMMENTED** [review]
Some comments:
1. `ClientCredentialsAuth` appears in several places.  Maybe this should be abstracted out to a helper that does auth?  One example from iceprod: https://github.com/WIPACrepo/iceprod/blob/master/iceprod/client_auth.py
2. If this is run as a user, it should be using `SavedDeviceGrantAuth`.
3. The client might be better as "file-catalog-indexer" - though note that if you intend to support both `ClientCredentialsAuth` and `SavedDeviceGrantAuth` they will need separate Keycloak clients.

As for IceProd integration, that could use this same token by adding an additional audience in Keycloak.

File: indexer/index.py [hunk 1]
```diff
@@ -23,11 +23,13 @@
...
+    """TypedDict for ClientCredentialsAuth parameters."""
 
-    url: str
-    token: str
+    address: str
+    token_url: str
+    client_id: str
+    client_secret: str
     timeout: int
     retries: int
```
Comment (ric-evans) [inline indexer/index.py:34]: Reminds me of when I was really into `TypedDict` for decreasing # of args... :smile: No changes needed

File: tests/integration/resources/test_path_collector.py
Comment (ric-evans) [inline tests/integration/resources/test_path_collector.py:1]: Looks much better now :+1:

**Review by ric-evans — COMMENTED** [review]
Looks good to me. Nice job fixing all those imports. Feel free to re-request me after future updates

File: resources/find_nonindexed/get_nonindexed_files.py [hunk 2]
```diff
@@ -1,22 +1,25 @@
...
-    rc = RestClient(
-        "https://file-catalog.icecube.wisc.edu/",
-        token=token,
+    rc = ClientCredentialsAuth(
+        address="https://file-catalog.icecube.wisc.edu/",
+        token_url="https://keycloak.icecube.wisc.edu/auth/realms/IceCube",
+        client_id="file-catalog",
+        client_secret=client_secret,
         timeout=60 * 60,  # 1 hour
         retries=24,  # 1 day
```
Comment (dsschult) [inline resources/find_nonindexed/get_nonindexed_files.py:24]: The timeout and retries seem exceedingly large?  I know of no cases where you'd want to sit there for a day instead of failing (and trying to run this again later).

File: indexer/defaults.py [hunk 3]
```diff
@@ -8,7 +8,7 @@
 PATHS_FILE = ""
 BLACKLIST = None
 BLACKLIST_FILE = ""
-URL = "https://file-catalog.icecube.wisc.edu/"
+ADDRESS = "https://file-catalog.icecube.wisc.edu/"
 TIMEOUT = 30  # seconds
 RETRIES = int((60 / TIMEOUT) * _AGGREGATE_LATENCY_MINUTES)
```
Comment (dsschult) [inline indexer/defaults.py:13]: Again with the high timeout and large number of retries.  Is there a reasoning?

File: resources/find_nonindexed/get_nonindexed_files.py (hunk 2 above)
Comment (ric-evans) [reply resources/find_nonindexed/get_nonindexed_files.py:24]: I wrote this part and cannot remember the reasoning. It could've been a misunderstanding of how those two arguments work--hard to say. It'd probably be best to require these as CL args

File: indexer/defaults.py (hunk 3 above)
Comment (ric-evans) [reply indexer/defaults.py:13]: https://github.com/WIPACrepo/file-catalog-indexer/pull/60#discussion_r1137610718, I recommend CL ags

File: indexer/client_auth.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            client_secret=args.oauth_client_secret,
+            timeout=args.rest_timeout,
+            retries=args.rest_retries,
+        )
+    else:
+        # raise an error if we can't honor the user's request
+        if args.rest_retries != defaults.REST_RETRIES:
+            raise ValueError("user credentials do not support custom REST_RETRIES or --rest-retries")
+        if args.rest_timeout != defaults.REST_TIMEOUT:
+            raise ValueError("user credentials do not support custom REST_TIMEOUT or --rest-timeout")
```
Comment (dsschult) [inline indexer/client_auth.py:83]: I suppose we could fix rest-tools to allow passing the retries and timeout. That's probably a bug.

File: indexer/client_auth.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+        # otherwise, let's make that RestClient
+        LOG.debug('Using user credentials to authenticate with IceProd')
+        if args.oauth_client_id == 'file-catalog-indexer':
+            args.oauth_client_id = 'file-catalog-indexer-public'
+        return SavedDeviceGrantAuth(
+            address=args.iceprod_rest_url,
+            filename='.iceprod-auth',
```
Comment (dsschult) [inline indexer/client_auth.py:122]: I think you want this to use the same file as the file catalog auth, or you'll need to log in a second time.  If you're ok logging in a second time, then the client_id should be `iceprod-public` to separate things out.

File: indexer/index.py [hunk 6]
```diff
@@ -308,68 +282,60 @@ def validate_path(path: str) -> None:
...
-    patch: bool = defaults.PATCH,
-    iceprodv2_rc_token: str = defaults.ICEPRODV2_RC_TOKEN,
-    iceprodv1_db_pass: str = defaults.ICEPRODV1_DB_PASS,
-    dryrun: bool = defaults.DRYRUN,
-    non_recursive: bool = defaults.NON_RECURSIVE,
-    n_processes: int = defaults.N_PROCESSES,
-    token_url: str = defaults.TOKEN_URL,
-    client_id: str = defaults.CLIENT_ID,
-) -> None:
+def index(args: Namespace) -> None:
```
Comment (ric-evans) [inline indexer/index.py:285]: This removes the feasibility to import and call `index()`, see https://github.com/WIPACrepo/file-catalog-indexer#from-indexerindex-import-index. It cleans the code up nicely, but it's something worth deciding now on how we're going to use the tool. You're the expert, so if we're not using the importability, let's nix it and update the README.

File: indexer/metadata/simulation/iceprod_tools.py [hunk 7]
```diff
@@ -1,10 +1,12 @@
+# iceprod_tools.py
```
Comment (ric-evans) [inline indexer/metadata/simulation/iceprod_tools.py:1]: Curious, why is this here?

**Review by ric-evans — CHANGES_REQUESTED** [review]

File: indexer/client_auth.py (hunk 4 above)
Comment (blinkdog) [reply indexer/client_auth.py:83]: I opened a PR on rest-tools to add support for this.

File: indexer/index.py (hunk 6 above)
Comment (blinkdog) [reply indexer/index.py:285]: I'll make some changes to preserve importability.
My specific use cases (lta and retro-disk-copy) use it through a subprocess rather than directly.
However, that doesn't mean somebody won't want to in the future.

File: indexer/index.py [hunk 8]
```diff
@@ -438,7 +401,7 @@ def index(
     parser.add_argument(
         "-n",
         "--non-recursive",
-        default=False,
+        default=defaults.NON_RECURSIVE,
         action="store_true",
         help="do not recursively index / do not descend into subdirectories",
     )
```
[the remark this answers is not in the document]
Comment (blinkdog) [reply indexer/index.py:407]: Good catch. I didn't realize that I had loaded that foot gun.

I'll keep the defaults as `False` and just remove the constants altogether. Considering the way these things propagate down from the command line, it'll be awfully confusing if we do decide to update the constant and nothing seems to be different because we've been using them in a way that ignores them anyway.

File: indexer/metadata/simulation/iceprod_tools.py (hunk 7 above)
Comment (blinkdog) [reply indexer/metadata/simulation/iceprod_tools.py:1]: Just an old habit. I like to put the name of the file in a comment at the top.
I think it comes from adding the GPL stanza. :smile:

File: indexer/client_auth.py [hunk 9]
```diff
@@ -63,63 +63,70 @@ def add_auth_to_argparse(parser: ArgumentParser) -> None:
...
+    if oauth_config["oauth_client_secret"]:
         LOG.debug('Using client credentials to authenticate with IceProd')
         return ClientCredentialsAuth(
-            address=args.iceprod_rest_url,
-            token_url=args.oauth_url,
-            client_id=args.oauth_client_id,
-            client_secret=args.oauth_client_secret,
-            timeout=args.rest_timeout,
-            retries=args.rest_retries,
+            address=rest_config["file_catalog_rest_url"],
```
Comment (dsschult) [inline indexer/client_auth.py:116]: This should be the iceprod rest url.

File: indexer/client_auth.py [hunk 10]
```diff
@@ -63,63 +63,70 @@ def add_auth_to_argparse(parser: ArgumentParser) -> None:
...
-        # raise an error if we can't honor the user's request
-        if args.rest_retries != defaults.REST_RETRIES:
-            raise ValueError("user credentials do not support custom REST_RETRIES or --rest-retries")
-        if args.rest_timeout != defaults.REST_TIMEOUT:
-            raise ValueError("user credentials do not support custom REST_TIMEOUT or --rest-timeout")
-
-        # otherwise, let's make that RestClient
         LOG.debug('Using user credentials to authenticate with the File Catalog')
-        if args.oauth_client_id == 'file-catalog-indexer':
-            args.oauth_client_id = 'file-catalog-indexer-public'
```
Comment (dsschult) [inline indexer/client_auth.py]: You still need these lines somewhere, since the user client needs to be different than the credentials client.

I guess you could put them in `create_oauth_config` if you want.

**Review by dsschult — CHANGES_REQUESTED** [review]

File: indexer/config.py
Comment (ric-evans) [inline indexer/config.py:1]: I'm partial to dataclasses currently, but that's an opinion, not a critique :smile:

**Review by ric-evans — COMMENTED** [review]
A few things, but no dealbreakers. I'm leaving the auth reviewing to David.

**Comment by ric-evans** [comment]
Forgot to mention, since these updates are breaking, let's make a major release (`[major]`).

**Comment by blinkdog** [comment]
> Forgot to mention, since these updates are breaking, let's make a major release (`[major]`).

The bot triggers on that tag in the title of the PR right?

File: indexer/client_auth.py (hunk 9 above)
Comment (blinkdog) [reply indexer/client_auth.py]: Ahhh, great catch. That would have been confusing.

File: indexer/client_auth.py (hunk 10 above)
Comment (blinkdog) [reply indexer/client_auth.py]: I've added some lines to `create_oauth_config` for this.

File: indexer/generate.py [hunk 11]
```diff
@@ -20,28 +28,34 @@ def main() -> None:
...
     parser.add_argument("-l", "--log", default="INFO", help="the output logging level")
-
+    parser.add_argument("-s", "--site", required=True, help='site value of the "locations" object')
+    add_auth_to_argparse(parser)
     args = parser.parse_args()
-    coloredlogs.install(level=args.log)
+
+    coloredlogs.install(level=args.log.upper())
     for arg, val in vars(args).items():
         logging.warning(f"{arg}: {val}")
```
[the remark this answers is not in the document]
Comment (blinkdog) [reply indexer/generate.py]: I modified this to use logging_tools.
Check the calls to make sure I didn't misuse the API?

**Review by ric-evans — APPROVED** [review]

**Comment by ric-evans** [comment]
> > Forgot to mention, since these updates are breaking, let's make a major release (`[major]`).
> 
> The bot triggers on that tag in the title of the PR right?

Yup!
https://github.com/WIPACrepo/file-catalog-indexer/blob/new-auth-file-catalog/setup.cfg#L52-L54

**Comment by dsschult** [comment]
> > > Forgot to mention, since these updates are breaking, let's make a major release (`[major]`).
> > 
> > 
> > The bot triggers on that tag in the title of the PR right?
> 
> Yup! https://github.com/WIPACrepo/file-catalog-indexer/blob/new-auth-file-catalog/setup.cfg#L52-L54

Technically, it's in whatever the merge commit to `master`  is.  But if you "Squash and merge" it uses the PR title by default.

**Review by dsschult — APPROVED** [review]
